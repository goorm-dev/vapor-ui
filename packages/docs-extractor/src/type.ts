import * as fs from 'fs';
import * as path from 'path';
import * as ts from 'typescript';

import { ComponentAnalyzer } from './component';
import { ExternalTypeResolver } from './external';
import { PropsAnalyzer } from './props';
import type { ComponentTypeInfo, TypeExtractorConfig } from './types/types';
import { getJSDocDescription, isReactReturnType } from './utils';
import { VanillaExtractAnalyzer } from './vanilla-extract';

/**
 * Refactored TypeExtractor with separated concerns and modular architecture
 *
 * This class orchestrates the various analyzers to extract comprehensive
 * component type information including props, default values, and Base UI integration
 */
export class TypeExtractor {
    private program!: ts.Program;
    private checker!: ts.TypeChecker;
    private config: TypeExtractorConfig;
    // Specialized analyzers
    private vanillaExtractAnalyzer!: VanillaExtractAnalyzer;
    private componentAnalyzer!: ComponentAnalyzer;
    private propsAnalyzer!: PropsAnalyzer;
    private externalTypeResolver!: ExternalTypeResolver;

    constructor(config: TypeExtractorConfig) {
        this.config = config;
        this.initializeTypeScriptProgram();
        this.initializeAnalyzers();
    }

    /**
     * Extracts component type information from a file
     */
    extractComponentTypes(filePath: string): ComponentTypeInfo[] {
        const sourceFilePath = path.resolve(filePath);

        const sourceFile = this.program.getSourceFile(sourceFilePath);

        if (!sourceFile) {
            throw new Error(`Cannot find file: ${sourceFilePath}`);
        }

        const components: ComponentTypeInfo[] = [];

        // Get all exported symbols
        const moduleSymbol = this.checker.getSymbolAtLocation(sourceFile);
        if (moduleSymbol) {
            const exports = this.checker.getExportsOfModule(moduleSymbol);

            for (const exportSymbol of exports) {
                const exportResults = this.parseExport(exportSymbol, sourceFile);
                if (exportResults) {
                    components.push(...exportResults);
                }
            }
        }

        return components;
    }

    /**
     * Initializes the TypeScript program with configuration
     */
    private initializeTypeScriptProgram(): void {
        const resolvedConfigPath = path.resolve(this.config.configPath);
        const projectDirectory = path.dirname(resolvedConfigPath);

        const { config, error } = ts.readConfigFile(this.config.configPath, (filePath) =>
            fs.readFileSync(filePath).toString(),
        );

        if (error) {
            throw error;
        }

        const { options, errors, fileNames } = ts.parseJsonConfigFileContent(
            config,
            ts.sys,
            projectDirectory,
        );

        if (errors.length > 0) throw errors[0];

        // Include additional files if specified

        // Include external type files (Base UI, React types, etc.)
        this.externalTypeResolver = new ExternalTypeResolver(
            this.config.projectRoot || projectDirectory,
        );

        const externalTypeFiles = this.externalTypeResolver.resolveExternalTypeFiles(
            this.config.externalTypePaths,
        );

        this.program = ts.createProgram([...fileNames, ...externalTypeFiles], options);
        this.checker = this.program.getTypeChecker();
    }

    /**
     * Initializes specialized analyzer instances
     */
    private initializeAnalyzers(): void {
        this.vanillaExtractAnalyzer = new VanillaExtractAnalyzer(this.program);
        this.componentAnalyzer = new ComponentAnalyzer();
        this.propsAnalyzer = new PropsAnalyzer(this.checker, this.vanillaExtractAnalyzer);
    }

    /**
     * Parses an exported symbol to extract component information
     */
    private parseExport(
        exportSymbol: ts.Symbol,
        sourceFile: ts.SourceFile,
    ): ComponentTypeInfo[] | undefined {
        try {
            const exportDeclaration = exportSymbol.declarations?.[0];
            if (!exportDeclaration) {
                return;
            }

            if (ts.isExportSpecifier(exportDeclaration)) {
                // export { Button } pattern
                if (
                    ts.isExportDeclaration(exportDeclaration.parent.parent) &&
                    exportDeclaration.parent.parent.moduleSpecifier !== undefined
                ) {
                    // Skip re-exports e.g., export { Button } from './Button';
                    return;
                }

                const targetSymbol =
                    this.checker.getExportSpecifierLocalTargetSymbol(exportDeclaration);
                if (!targetSymbol) {
                    return;
                }

                let type: ts.Type;
                if (targetSymbol.declarations?.length) {
                    type = this.checker.getTypeAtLocation(targetSymbol.declarations[0]);
                } else {
                    type = this.checker.getTypeOfSymbol(targetSymbol);
                }

                return this.createComponentInfo(exportSymbol.name, targetSymbol, type, sourceFile);
            }
        } catch (error) {
            console.error('Export 파싱 중 오류:', error);
        }
    }

    /**
     * Creates component type information from symbol and type data
     */
    private createComponentInfo(
        name: string,
        symbol: ts.Symbol,
        type: ts.Type,
        sourceFile: ts.SourceFile,
    ): ComponentTypeInfo[] | undefined {
        // Check if it's a React component
        if (!isReactReturnType(type, this.checker)) {
            return;
        }

        const componentName = name;
        const description = getJSDocDescription(symbol, this.checker);

        // Extract props
        const propsType = this.propsAnalyzer.extractPropsType(type);
        const props = propsType ? this.propsAnalyzer.extractProps(propsType, sourceFile) : [];

        // Extract display name and default element
        const displayName = this.componentAnalyzer.extractDisplayName(symbol);
        const defaultElement = this.componentAnalyzer.extractDefaultElement(symbol);

        return [
            {
                name: componentName,
                displayName,
                description,
                props,
                defaultElement,
            },
        ];
    }
}

/**
 * Convenience function for extracting component types from a file
 */
export function extractComponentTypesFromFile(
    configPath: string,
    filePath: string,
    includeFiles?: string[],
    externalTypePaths?: string[],
): ComponentTypeInfo[] {
    const config: TypeExtractorConfig = {
        configPath,
        files: includeFiles,
        externalTypePaths,
    };

    const extractor = new TypeExtractor(config);
    return extractor.extractComponentTypes(filePath);
}
