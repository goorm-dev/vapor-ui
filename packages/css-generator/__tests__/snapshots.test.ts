import type { SemanticMappingConfig } from '@vapor-ui/color-generator';
import { describe, expect, it } from 'vitest';

import {
    generateColorCSS,
    generateCompleteCSS,
    generateRadiusCSS,
    generateScalingCSS,
} from '../src/index';
import type { CompleteCSSConfig } from '../src/types';

const MOCK_COLOR_CONFIG = {
    primary: { name: 'mint', color: '#6af574ff' },
    background: {
        name: 'neutral',
        color: '#F8FAFC',
        lightness: {
            light: 98,
            dark: 8,
        },
    },
} satisfies SemanticMappingConfig;

const MOCK_TOTAL_CONFIG: CompleteCSSConfig = {
    colors: MOCK_COLOR_CONFIG,
    scaling: 1.15,
    radius: 'lg',
};

describe('CSS Generator Snapshots', () => {
    /* -------------------------------------------------------------------------------------------------
     * Color CSS
     * -----------------------------------------------------------------------------------------------*/
    it('should match generateColorCSS snapshot', () => {
        const result = generateColorCSS(MOCK_COLOR_CONFIG);
        expect(result).toMatchSnapshot();
    });

    it('should match generateColorCSS with options snapshot', () => {
        const result = generateColorCSS(MOCK_COLOR_CONFIG, {
            prefix: 'custom',
            format: 'compact',
        });
        expect(result).toMatchSnapshot();
    });

    /* -------------------------------------------------------------------------------------------------
     * Radius CSS
     * -----------------------------------------------------------------------------------------------*/
    it('should match generateRadiusCSS snapshot', () => {
        const result = generateRadiusCSS('full');
        expect(result).toMatchSnapshot();
    });

    it('should match generateRadiusCSS with options snapshot', () => {
        const result = generateRadiusCSS('sm', {
            prefix: 'custom',
            format: 'compact',
        });
        expect(result).toMatchSnapshot();
    });

    /* -------------------------------------------------------------------------------------------------
     * Scaling CSS
     * -----------------------------------------------------------------------------------------------*/

    it('should match generateScalingCSS snapshot', () => {
        const result = generateScalingCSS(1.15);
        expect(result).toMatchSnapshot();
    });

    it('should match generateScalingCSS with options snapshot', () => {
        const result = generateScalingCSS(1.25, {
            prefix: 'custom',
            format: 'compact',
        });
        expect(result).toMatchSnapshot();
    });

    /* -------------------------------------------------------------------------------------------------
     * Complete CSS
     * -----------------------------------------------------------------------------------------------*/

    it('should match generateCompleteCSS snapshot', () => {
        const result = generateCompleteCSS(MOCK_TOTAL_CONFIG);
        expect(result).toMatchSnapshot();
    });

    it('should match generateCompleteCSS with options snapshot', () => {
        const result = generateCompleteCSS(MOCK_TOTAL_CONFIG, {
            includeColorComments: true,
            prefix: 'custom',
            format: 'compact',
        });
        expect(result).toMatchSnapshot();
    });
});
