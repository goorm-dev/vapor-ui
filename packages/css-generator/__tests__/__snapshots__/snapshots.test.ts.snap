// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CSS Generator Snapshots > should match generateColorCSS snapshot 1`] = `
":root, [data-vapor-theme="light"] {
    --vapor-color-background-canvas: #f8fafc;
    --vapor-color-mint-050: #cdfdcd;
    --vapor-color-mint-100: #6af574ff;
    --vapor-color-mint-200: #58db60;
    --vapor-color-mint-300: #3eb643;
    --vapor-color-mint-400: #33a636;
    --vapor-color-mint-500: #178618;
    --vapor-color-mint-600: #006a00;
    --vapor-color-mint-700: #005700;
    --vapor-color-mint-800: #004000;
    --vapor-color-mint-900: #002a00;
    --vapor-color-neutral-050: #eff1f3;
    --vapor-color-neutral-100: #dbddde;
    --vapor-color-neutral-200: #c0c2c3;
    --vapor-color-neutral-300: #9fa0a2;
    --vapor-color-neutral-400: #909193;
    --vapor-color-neutral-500: #737374;
    --vapor-color-neutral-600: #5a5b5b;
    --vapor-color-neutral-700: #494a4a;
    --vapor-color-neutral-800: #363637;
    --vapor-color-neutral-900: #232324;
    --vapor-color-background-primary-100: var(--vapor-color-mint-100);
    --vapor-color-background-primary-200: var(--vapor-color-mint-100);
    --vapor-color-foreground-primary-100: var(--vapor-color-mint-200);
    --vapor-color-foreground-primary-200: var(--vapor-color-mint-300);
    --vapor-color-border-primary: var(--vapor-color-mint-100);
    --vapor-color-button-foreground-primary: var(--vapor-color-black);
}

[data-vapor-theme="dark"] {
    --vapor-color-background-canvas: #161717;
    --vapor-color-mint-050: #002400;
    --vapor-color-mint-100: #003600;
    --vapor-color-mint-200: #004a00;
    --vapor-color-mint-300: #006700;
    --vapor-color-mint-400: #017401;
    --vapor-color-mint-500: #239325;
    --vapor-color-mint-600: #3bb13f;
    --vapor-color-mint-700: #4dcb53;
    --vapor-color-mint-800: #62e96b;
    --vapor-color-mint-900: #b8fbb9;
    --vapor-color-neutral-050: #1e1e1e;
    --vapor-color-neutral-100: #2d2d2d;
    --vapor-color-neutral-200: #3f3f40;
    --vapor-color-neutral-300: #575758;
    --vapor-color-neutral-400: #636364;
    --vapor-color-neutral-500: #7e7f81;
    --vapor-color-neutral-600: #9a9c9d;
    --vapor-color-neutral-700: #b1b2b4;
    --vapor-color-neutral-800: #cdcfd1;
    --vapor-color-neutral-900: #e9ebec;
    --vapor-color-background-primary-100: var(--vapor-color-mint-800);
    --vapor-color-background-primary-200: var(--vapor-color-mint-800);
    --vapor-color-foreground-primary-100: var(--vapor-color-mint-900);
    --vapor-color-foreground-primary-200: var(--vapor-color-mint-900);
    --vapor-color-border-primary: var(--vapor-color-mint-800);
    --vapor-color-button-foreground-primary: var(--vapor-color-black);
}"
`;

exports[`CSS Generator Snapshots > should match generateColorCSS with options snapshot 1`] = `":root, [data-vapor-theme="light"]{--custom-color-background-canvas:#f8fafc;--custom-color-mint-050:#cdfdcd;--custom-color-mint-100:#6af574ff;--custom-color-mint-200:#58db60;--custom-color-mint-300:#3eb643;--custom-color-mint-400:#33a636;--custom-color-mint-500:#178618;--custom-color-mint-600:#006a00;--custom-color-mint-700:#005700;--custom-color-mint-800:#004000;--custom-color-mint-900:#002a00;--custom-color-neutral-050:#eff1f3;--custom-color-neutral-100:#dbddde;--custom-color-neutral-200:#c0c2c3;--custom-color-neutral-300:#9fa0a2;--custom-color-neutral-400:#909193;--custom-color-neutral-500:#737374;--custom-color-neutral-600:#5a5b5b;--custom-color-neutral-700:#494a4a;--custom-color-neutral-800:#363637;--custom-color-neutral-900:#232324;--custom-color-background-primary-100:var(--custom-color-mint-100);--custom-color-background-primary-200:var(--custom-color-mint-100);--custom-color-foreground-primary-100:var(--custom-color-mint-200);--custom-color-foreground-primary-200:var(--custom-color-mint-300);--custom-color-border-primary:var(--custom-color-mint-100);--custom-color-button-foreground-primary:var(--custom-color-black)}[data-vapor-theme="dark"]{--custom-color-background-canvas:#161717;--custom-color-mint-050:#002400;--custom-color-mint-100:#003600;--custom-color-mint-200:#004a00;--custom-color-mint-300:#006700;--custom-color-mint-400:#017401;--custom-color-mint-500:#239325;--custom-color-mint-600:#3bb13f;--custom-color-mint-700:#4dcb53;--custom-color-mint-800:#62e96b;--custom-color-mint-900:#b8fbb9;--custom-color-neutral-050:#1e1e1e;--custom-color-neutral-100:#2d2d2d;--custom-color-neutral-200:#3f3f40;--custom-color-neutral-300:#575758;--custom-color-neutral-400:#636364;--custom-color-neutral-500:#7e7f81;--custom-color-neutral-600:#9a9c9d;--custom-color-neutral-700:#b1b2b4;--custom-color-neutral-800:#cdcfd1;--custom-color-neutral-900:#e9ebec;--custom-color-background-primary-100:var(--custom-color-mint-800);--custom-color-background-primary-200:var(--custom-color-mint-800);--custom-color-foreground-primary-100:var(--custom-color-mint-900);--custom-color-foreground-primary-200:var(--custom-color-mint-900);--custom-color-border-primary:var(--custom-color-mint-800);--custom-color-button-foreground-primary:var(--custom-color-black)}"`;

exports[`CSS Generator Snapshots > should match generateCompleteCSS snapshot 1`] = `
":root, [data-vapor-theme="light"] {
    --vapor-color-background-canvas: #f8fafc;
    --vapor-color-mint-050: #cdfdcd;
    --vapor-color-mint-100: #6af574ff;
    --vapor-color-mint-200: #58db60;
    --vapor-color-mint-300: #3eb643;
    --vapor-color-mint-400: #33a636;
    --vapor-color-mint-500: #178618;
    --vapor-color-mint-600: #006a00;
    --vapor-color-mint-700: #005700;
    --vapor-color-mint-800: #004000;
    --vapor-color-mint-900: #002a00;
    --vapor-color-neutral-050: #eff1f3;
    --vapor-color-neutral-100: #dbddde;
    --vapor-color-neutral-200: #c0c2c3;
    --vapor-color-neutral-300: #9fa0a2;
    --vapor-color-neutral-400: #909193;
    --vapor-color-neutral-500: #737374;
    --vapor-color-neutral-600: #5a5b5b;
    --vapor-color-neutral-700: #494a4a;
    --vapor-color-neutral-800: #363637;
    --vapor-color-neutral-900: #232324;
    --vapor-color-background-primary-100: var(--vapor-color-mint-100);
    --vapor-color-background-primary-200: var(--vapor-color-mint-100);
    --vapor-color-foreground-primary-100: var(--vapor-color-mint-200);
    --vapor-color-foreground-primary-200: var(--vapor-color-mint-300);
    --vapor-color-border-primary: var(--vapor-color-mint-100);
    --vapor-color-button-foreground-primary: var(--vapor-color-black);
    --vapor-scale-factor: 1.15;
    --vapor-radius-factor: 1.5;
}

[data-vapor-theme="dark"] {
    --vapor-color-background-canvas: #161717;
    --vapor-color-mint-050: #002400;
    --vapor-color-mint-100: #003600;
    --vapor-color-mint-200: #004a00;
    --vapor-color-mint-300: #006700;
    --vapor-color-mint-400: #017401;
    --vapor-color-mint-500: #239325;
    --vapor-color-mint-600: #3bb13f;
    --vapor-color-mint-700: #4dcb53;
    --vapor-color-mint-800: #62e96b;
    --vapor-color-mint-900: #b8fbb9;
    --vapor-color-neutral-050: #1e1e1e;
    --vapor-color-neutral-100: #2d2d2d;
    --vapor-color-neutral-200: #3f3f40;
    --vapor-color-neutral-300: #575758;
    --vapor-color-neutral-400: #636364;
    --vapor-color-neutral-500: #7e7f81;
    --vapor-color-neutral-600: #9a9c9d;
    --vapor-color-neutral-700: #b1b2b4;
    --vapor-color-neutral-800: #cdcfd1;
    --vapor-color-neutral-900: #e9ebec;
    --vapor-color-background-primary-100: var(--vapor-color-mint-800);
    --vapor-color-background-primary-200: var(--vapor-color-mint-800);
    --vapor-color-foreground-primary-100: var(--vapor-color-mint-900);
    --vapor-color-foreground-primary-200: var(--vapor-color-mint-900);
    --vapor-color-border-primary: var(--vapor-color-mint-800);
    --vapor-color-button-foreground-primary: var(--vapor-color-black);
}"
`;

exports[`CSS Generator Snapshots > should match generateCompleteCSS with options snapshot 1`] = `
"/*
 * Vapor UI Theme
 * Generated with @vapor-ui/css-generator
 * 
 * Colors: mint (#6af574ff)
 * Scaling: 1.15
 * Radius: lg
 */



:root,[data-vapor-theme="light"]{--custom-color-background-canvas:#f8fafc;--custom-color-mint-050:#cdfdcd;--custom-color-mint-100:#6af574ff;--custom-color-mint-200:#58db60;--custom-color-mint-300:#3eb643;--custom-color-mint-400:#33a636;--custom-color-mint-500:#178618;--custom-color-mint-600:#006a00;--custom-color-mint-700:#005700;--custom-color-mint-800:#004000;--custom-color-mint-900:#002a00;--custom-color-neutral-050:#eff1f3;--custom-color-neutral-100:#dbddde;--custom-color-neutral-200:#c0c2c3;--custom-color-neutral-300:#9fa0a2;--custom-color-neutral-400:#909193;--custom-color-neutral-500:#737374;--custom-color-neutral-600:#5a5b5b;--custom-color-neutral-700:#494a4a;--custom-color-neutral-800:#363637;--custom-color-neutral-900:#232324;--custom-color-background-primary-100:var(--custom-color-mint-100);--custom-color-background-primary-200:var(--custom-color-mint-100);--custom-color-foreground-primary-100:var(--custom-color-mint-200);--custom-color-foreground-primary-200:var(--custom-color-mint-300);--custom-color-border-primary:var(--custom-color-mint-100);--custom-color-button-foreground-primary:var(--custom-color-black);--custom-scale-factor:1.15;--custom-radius-factor:1.5}

[data-vapor-theme="dark"]{--custom-color-background-canvas:#161717;--custom-color-mint-050:#002400;--custom-color-mint-100:#003600;--custom-color-mint-200:#004a00;--custom-color-mint-300:#006700;--custom-color-mint-400:#017401;--custom-color-mint-500:#239325;--custom-color-mint-600:#3bb13f;--custom-color-mint-700:#4dcb53;--custom-color-mint-800:#62e96b;--custom-color-mint-900:#b8fbb9;--custom-color-neutral-050:#1e1e1e;--custom-color-neutral-100:#2d2d2d;--custom-color-neutral-200:#3f3f40;--custom-color-neutral-300:#575758;--custom-color-neutral-400:#636364;--custom-color-neutral-500:#7e7f81;--custom-color-neutral-600:#9a9c9d;--custom-color-neutral-700:#b1b2b4;--custom-color-neutral-800:#cdcfd1;--custom-color-neutral-900:#e9ebec;--custom-color-background-primary-100:var(--custom-color-mint-800);--custom-color-background-primary-200:var(--custom-color-mint-800);--custom-color-foreground-primary-100:var(--custom-color-mint-900);--custom-color-foreground-primary-200:var(--custom-color-mint-900);--custom-color-border-primary:var(--custom-color-mint-800);--custom-color-button-foreground-primary:var(--custom-color-black)}"
`;

exports[`CSS Generator Snapshots > should match generateRadiusCSS snapshot 1`] = `
":root, [data-vapor-theme="light"] {
    --vapor-radius-factor: 3;
}"
`;

exports[`CSS Generator Snapshots > should match generateRadiusCSS with options snapshot 1`] = `":root, [data-vapor-theme="light"]{--custom-radius-factor:0.5}"`;

exports[`CSS Generator Snapshots > should match generateScalingCSS snapshot 1`] = `
":root, [data-vapor-theme="light"] {
    --vapor-scale-factor: 1.15;
}"
`;

exports[`CSS Generator Snapshots > should match generateScalingCSS with options snapshot 1`] = `":root, [data-vapor-theme="light"]{--custom-scale-factor:1.25}"`;